File: 20221019/trainingBook/src/main/java/TrainingBook.java
import java.util.Arrays;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

public class TrainingBook {
    public int solution(int n, int[] lost, int[] reserve) {
        int answer = 0;

        Arrays.sort(lost);
        Arrays.sort(reserve);

        Map<Integer, Integer> trainingBook = new LinkedHashMap<>();

        List<Integer> reserved = getIntegers(reserve, lost);

        List<Integer> losted = getIntegers(lost, reserve);

        for (int i = 0; i < n; i += 1) {
            if (reserved.contains(i + 1)) {
                trainingBook.put(i + 1, 2);
                continue;
            }

            if (losted.contains(i + 1)) {
                trainingBook.put(i + 1, 0);
                continue;
            }

            trainingBook.put(i + 1, 1);
        }

        for (int person : reserved) {
            if (person != 1 && trainingBook.get(person - 1) == 0) {
                trainingBook.replace(person - 1, 1);
                trainingBook.replace(person, 1);
                continue;
            }

            if (person != n && trainingBook.get(person + 1) == 0) {
                trainingBook.replace(person + 1, 1);
                trainingBook.replace(person, 1);
            }
        }

        for (int number : trainingBook.values()) {
            if (number != 0) {
                answer += 1;
            }
        }

        return answer;
    }

    public List<Integer> getIntegers(int[] lost, int[] reserve) {
        List<Integer> losted =
                Arrays.stream(lost)
                .filter(one -> Arrays.stream(reserve).noneMatch(r -> r == one))
                .boxed().collect(Collectors.toList());
        return losted;
    }
}

File: 20221020/trainingBook.js
export default function trainingBook(n, lost, reserve) {
  const realLost = [...lost].sort().filter((person) => !reserve.includes(person));
  const realReserve = [...reserve].sort().filter((person) => !lost.includes(person));

  const realTrainingBook = [...Array(n)].map((_, i) => {
    if (realLost.includes(i + 1)) {
      return 0;
    }
    if (realReserve.includes(i + 1)) {
      return 2;
    }
    return 1;
  });

  realReserve.forEach((reserve) => {
    const index = reserve - 1;
    if (realTrainingBook[index - 1] === 0 && reserve !== 1) {
      realTrainingBook[index] = 1;
      realTrainingBook[index - 1] = 1;
      return;
    }

    if (realTrainingBook[index + 1] === 0 && reserve !== n) {
      realTrainingBook[index] = 1;
      realTrainingBook[index + 1] = 1;
    }
  });

  return realTrainingBook.filter((person) => person !== 0).length;
}
